enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/upstream/strategies/TypicallyConnectToRandomReadReplicaStrategy.java

```diff
@@ -54,18 +62,15 @@ public TypicallyConnectToRandomReadReplicaStrategy( int connectToCoreInterval )
         }
         else
         {
-            Optional<MemberId> memberId = getReadReplicaMemberId();
-            if ( !memberId.isPresent() )
-            {
-                memberId = getCoreMemberId();
-            }
-            return memberId;
-        }
-    }
+            // shuffled members
+            List<MemberId> readReplicaMembers = new ArrayList<>( topologyService.localReadReplicas().members().keySet() );
+            Collections.shuffle( readReplicaMembers );
 
-    private Optional<MemberId> getReadReplicaMemberId()
-    {
-        return topologyService.localReadReplicas().randomReadReplicaMemberId();
+            List<MemberId> coreMembers = new ArrayList<>( topologyService.localCoreServers().members().keySet() );
+            Collections.shuffle( coreMembers );
+
+            return Stream.concat( readReplicaMembers.stream(), coreMembers.stream() ).filter( not( myself::equals ) ).findFirst();
+        }
     }
 
     private Optional<MemberId> getCoreMemberId()
```

Comment:
This can still return self

Reply:
Yes, however that gets passed into a concatenated stream and then self is filtered from it. So ultimately it does get self removed.

Reply:
getCoreMemberId is returned directly is it not?

Reply:
Yes, I missed that thanks - fixed